libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,779 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  handle FTP operations over MAVLink
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "GCS.h"
+#include <AP_Math/crc.h>
+
+#if HAL_WITH_MAVFTP
+
+#if HAL_OS_POSIX_IO
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <stdio.h>
+#include <time.h>
+#include <dirent.h>
+#if defined(__APPLE__) && defined(__MACH__)
+#include <sys/param.h>
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#endif
+#endif
+
+#if HAL_OS_FATFS_IO
+#include <stdio.h>
+#endif
+
+
+extern const AP_HAL::HAL& hal;
+
+static const char	kDirentFile = 'F';	///< Identifies File returned from List command
+static const char	kDirentDir = 'D';	///< Identifies Directory returned from List command
+static const char	kDirentSkip = 'S';	///< Identifies Skipped entry from List command
+#if 0
+#define debug(str) gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), str)
+#else
+#define debug(str)
+#endif
+
+void GCS_MAVLINK::_ftp_init()
+{
+    strcpy(_ftp_current_abs_path, HAL_BOARD_STORAGE_DIRECTORY);
+}
+
+char *GCS_MAVLINK::_data_as_cstring(FTPPayloadHeader *payload)
+{
+    // guarantee null termination
+    if (payload->size < MAX_FTP_DATA_LENGTH) {
+        ((char*)payload->data)[payload->size] = '\0';
+
+    } else {
+        //Discard the last character to ensure null termination
+        payload->data[MAX_FTP_DATA_LENGTH - 1] = '\0';
+    }
+
+    // and return data
+    return (char *) & (payload->data[0]);
+}
+
+bool GCS_MAVLINK::_set_current_abs_path(char* rel_path)
+{
+    if (_ftp_current_abs_path != nullptr) {
+        delete _ftp_current_abs_path;
+    }
+    _ftp_current_abs_path = new char[strlen(HAL_BOARD_STORAGE_DIRECTORY) + strlen(rel_path) + 2];
+    if (_ftp_current_abs_path == nullptr) {
+        return false;
+    }
+    strcpy(_ftp_current_abs_path, HAL_BOARD_STORAGE_DIRECTORY);
+    if (!(rel_path[0] == '/' && rel_path[1] == '\0')) {
+        strcpy(&_ftp_current_abs_path[strlen(HAL_BOARD_STORAGE_DIRECTORY)], rel_path);
+    }
+    return true;
+}
+
+void GCS_MAVLINK::handle_ftp_messages(mavlink_message_t *msg)
+{
+    if (!_ftp_timer_registered) {
+        //First FTP command received, register MAV FTP worker method
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&GCS_MAVLINK::_ftp_timer, void));
+        _ftp_timer_registered = true;
+    }
+    if (_ftp_sem.take_nonblocking()) {  //Do a nonblocking semaphore take, to ensure ftp worker is not active
+        //Update ftp_request structure
```

Comment:
Shouldn't you only do this if you've actually managed to complete the previous request that was pending? Otherwise we just drop the packet without any NAck?. If we moved to a queue of messages to handle then this would be much less of a problem.